### PR TITLE
Curved labels pt2

### PR DESCRIPTION
**Pull request by hjanetzek**
Follow-up PR's
- [ ] Handle line offset for curved labels.

### CODE REVIEW

**Comment by hjanetzek** [comment]
@karimnaaji Ready for testing and initial review. Please ignore the commented out debugging logs :)

**Comment by karimnaaji** [comment]
Starting to review this tomorrow!

File: core/src/style/textStyleBuilder.cpp [hunk 1]
```diff
@@ -209,59 +207,240 @@ bool TextStyleBuilder::addFeature(const Feature& _feat, const DrawRule& _rule) {
...
-                    if (d > tolerance) { break; }
+        for (size_t j = i + 1; j < _line.size()-1; j++) {
+            glm::vec2 dir2 = sampler.segmentDirection(j);
 
-                    // Skip merged segment
-                    merged = true;
-                    i += 1;
+            float angle = crossProduct(dir1, dir2);
+
+            if (std::abs(angle) > flipTolerance) {
```
Comment (karimnaaji) [inline core/src/style/textStyleBuilder.cpp]: Should we also have a similar rule on flip tolerance at screen space time? Such that we ensure text is in a readable direction. See _Place Denfert-Rochereau_ on the screenshot:

screen shot 2017-02-02 at 4 04 37 pm

File: core/src/labels/curvedLabel.cpp [hunk 2]
```diff
@@ -0,0 +1,264 @@
...
+
+        float currLength = sampler.point(i).z;
+        float segmentLength = currLength - prevLength;
+        count++;
+
+        if (start + width > currLength) {
+            p2 = glm::vec2(sampler.point(i));
+
+            rotation = sampler.segmentDirection(i-1);
+            _obbs.append({(p1 + p2) * 0.5f, rotation, segmentLength, dim.y});
```
Comment (karimnaaji) [inline core/src/labels/curvedLabel.cpp]: I think that at this point `p1` and `p2` are missing the label offset, what do you think of:

```
glm::vec2 obbPosition = (p1 + p2) * 0.5f;
glm::vec2 dir = sampler.segmentDirection(i);
glm::vec2 normal = glm::normalize(glm::vec2(-dir.y, dir.x));
obbPosition += normal * m_options.offset;
_obbs.append({obbPosition, rotation, segmentLength, dim.y});
```

To move the obbs along the normal defined by the segment they are currently attached to.

File: core/src/labels/curvedLabel.h [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+        return m_prio;
+    }
+
+    glm::vec2 modelCenter() const override {
+        return m_worldTransform[m_anchorPoint];
+    }
+
+protected:
+
+    const std::vector<glm::vec2> m_worldTransform;
```
Comment (karimnaaji) [inline core/src/labels/curvedLabel.h]: Should we name that `m_modelTransform` since we moved away from calling it world in previous PR?

File: core/src/style/textStyleBuilder.cpp (hunk 1 above)
Comment (nvkelso) [inline core/src/style/textStyleBuilder.cpp]: Yes, please :)

Comment (hjanetzek) [inline core/src/style/textStyleBuilder.cpp]: @karimnaaji the case above could be avoided in screen space placement by not allowing quads being placed from right-to-left (with some threshold). Though to me the above label is still within in readable threshold.

`flipTolerance` is checked to count the direction changes of the curve.

File: core/src/labels/curvedLabel.h (hunk 3 above)
Comment (hjanetzek) [inline core/src/labels/curvedLabel.h]: yup

File: core/src/labels/curvedLabel.cpp (hunk 2 above)
Comment (hjanetzek) [inline core/src/labels/curvedLabel.cpp]: Label offset for curved labels needs the be applied to the polyline in updateScreenTransform. I'll do this in the next PR.

Comment (karimnaaji) [inline core/src/labels/curvedLabel.cpp]: Ok sounds good to be in a separate PR.

File: core/src/labels/curvedLabel.cpp [hunk 4]
```diff
@@ -0,0 +1,266 @@
...
+#include "log.h"
+#include "style/textStyle.h"
+#include "text/fontContext.h"
+#include "textLabels.h"
+#include "util/geom.h"
+#include "util/lineSampler.h"
+#include "view/view.h"
+
+#include "alfons/path/splinePath.h"
+#include "alfons/path/ASPC.h"
```
Comment (karimnaaji) [inline core/src/labels/curvedLabel.cpp]: Both of those include seem unused

File: core/src/style/textStyleBuilder.cpp [hunk 5]
```diff
@@ -209,59 +207,240 @@ bool TextStyleBuilder::addFeature(const Feature& _feat, const DrawRule& _rule) {
...
 
-                    float d = sqPointSegmentDistance(pp, p1, p);
-                    if (d > tolerance) { break; }
+        for (size_t j = i + 1; j < _line.size()-1; j++) {
+            glm::vec2 dir2 = sampler.segmentDirection(j);
 
-                    // Skip merged segment
-                    merged = true;
-                    i += 1;
+            float angle = crossProduct(dir1, dir2);
```
Comment (karimnaaji) [inline core/src/style/textStyleBuilder.cpp]: Can we call this area instead? I think it's misleading to call it an angle.

File: core/src/style/textStyleBuilder.cpp (hunk 1 above)
Comment (karimnaaji) [inline core/src/style/textStyleBuilder.cpp]: Yep the rule for 'squiggly' labels is definitely important. The one up there seems to be _almost_ readable, but we can still run into cases where the letters can be upside down. It would be safe to add a rule to ensure that no quads gets upside down in screen space. I will give a look to implement that today.

Cases like this one are quite hard to read:
screen shot 2017-02-02 at 3 01 19 pm

Comment (hjanetzek) [inline core/src/style/textStyleBuilder.cpp]: Yeah, that one should be handled. You can probably check in `updateScreenTransform` that within the label range `sampler.segmentDirection(i).x > threshold`. (After having determined the overall label direction)

File: core/src/style/textStyleBuilder.cpp (hunk 5 above)
Comment (hjanetzek) [inline core/src/style/textStyleBuilder.cpp]: I've updated the code and docs - are you ok with angle?

https://www.physicsforums.com/threads/using-cross-product-to-find-angle-between-two-vectors.510814/

File: core/src/labels/curvedLabel.cpp (hunk 4 above)
Comment (hjanetzek) [inline core/src/labels/curvedLabel.cpp]: fixed

File: core/src/style/textStyleBuilder.cpp (hunk 5 above)
Comment (karimnaaji) [inline core/src/style/textStyleBuilder.cpp]: Yeah if dir1 and dir2 are normalized that's fine.

**Comment by karimnaaji** [comment]
Merging this after https://github.com/tangrams/tangram-es/pull/1263 so we do a patch release for iOS.

**Comment by tallytalwar** [comment]
Can we check all the places being tested for tangram JS, refer: https://github.com/tangrams/tangram/pull/484

**Comment by hjanetzek** [comment]
@tallytalwar  yes, we should ask our cartography team for final acceptance test after the next PR :)

File: core/src/style/textStyleBuilder.cpp [hunk 6]
```diff
@@ -322,9 +501,7 @@ TextStyle::Parameters TextStyleBuilder::applyRule(const DrawRule& _rule,
             p.labelOptions.priority = (float)priority;
         }
         _rule.get(StyleParamKey::text_collide, p.labelOptions.collide);
-        if (!_rule.get(StyleParamKey::text_interactive, p.interactive)) {
-            _rule.get(StyleParamKey::interactive, p.interactive);
-        }
+        _rule.get(StyleParamKey::text_interactive, p.interactive);
```
Comment (tallytalwar) [inline core/src/style/textStyleBuilder.cpp]: `interactive` parameter is inherited from the parent `point` style, if `text` style itself is not having one defined.

File: core/src/style/textStyleBuilder.cpp [hunk 7]
```diff
@@ -210,59 +208,240 @@ bool TextStyleBuilder::addFeature(const Feature& _feat, const DrawRule& _rule) {
...
-                    addLabel(_params, Label::Type::line, { a, a+b }, _rule);
-                    a += b;
+            if (!splitLine) {
+                // Go back within window to check for hard direction changes
+                for (int k = j - 1; k >= int(i); k--){
+                    if (glm::length2(sampler.segmentDirection(k) + dir2) < sqDirLimit) {
+                        //LOG("back break %s, %d %d", _params.text.c_str(), j, k);
+                        splitLine = true;
+                    }
+                    if (sampler.point(k).z < sampler.point(j).z - sampleWindow) {
```
Comment (tallytalwar) [inline core/src/style/textStyleBuilder.cpp]: @hjanetzek can you explain the use of `sampleWindow` here? I am not sure I get it or maybe rename it to something which explains the context in which its used.

Comment (hjanetzek) [inline core/src/style/textStyleBuilder.cpp]: The idea is to break the line also on sharp turns that are made of multiple segments. E.g. a 90° turn could be made of 3 30° 1px long segments. To detect these turns the current segment direction is compared to the previous segment directions within sampleWindow range.

File: core/src/style/textStyleBuilder.cpp (hunk 6 above)
Comment (hjanetzek) [inline core/src/style/textStyleBuilder.cpp]: Oh yes - fixed in pt3
